crates/daphne/src/pine/flp.rs

```diff
@@ -486,6 +471,106 @@ impl<F: FftFriendlyFieldElement> Type for PineType<F> {
     }
 }
 
+/// FLP used to check that the squared norm of the gradient is equal to the claimed value.
+///
+/// To ensure a gradient is valid, it is also necessary to run [`PineType`] on the encoded
+/// measurement.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct PineTypeSquaredNormEqual<F> {
+    pub(crate) cfg: PineConfig<F>,
+}
+
+impl<F: FftFriendlyFieldElement> Type for PineTypeSquaredNormEqual<F> {
+    type Measurement = (); // Not used by this type.
+    type AggregateResult = (); // Not used by this type.
+    type Field = F;
+
+    fn encode_measurement(&self, _measurement: &()) -> Result<Vec<F>, FlpError> {
+        unimplemented!("use PineType to encode the measurement")
+    }
+
+    fn decode_result(&self, _data: &[F], _num_measurements: usize) -> Result<(), FlpError> {
+        unimplemented!("use PineType to decode the aggregate result")
+    }
+
+    fn truncate(&self, _input: Vec<F>) -> Result<Vec<F>, FlpError> {
+        unimplemented!("use PineType to truncate the input")
+    }
+
+    fn output_len(&self) -> usize {
+        unimplemented!("use PineType to truncate the input")
+    }
+
+    fn gadget(&self) -> Vec<Box<dyn Gadget<F>>> {
+        vec![Box::new(ParallelSum::new(
+            Mul::new(self.cfg.gadget_calls_sq_norm_equal),
+            self.cfg.chunk_len_sq_norm_equal,
+        ))]
+    }
+
+    fn valid(
+        &self,
+        gadget: &mut Vec<Box<dyn Gadget<F>>>,
+        meas: &[F],
+        _joint_rand: &[F],
+        _num_shares: usize,
+    ) -> Result<F, FlpError> {
+        let mut buf = Vec::with_capacity(self.cfg.chunk_len_sq_norm_equal * 2);
+        let (encoded_gradient, rest) = meas.split_at(self.cfg.dimension);
+        let (sq_norm_v_bits, _rest) = rest.split_at(self.cfg.sq_norm_bits);
+
+        // Compute the squared norm of the gradient.
+        //
+        // This corresponds to the third call to `parallel_sum()` in the validity circuit of the
+        // reference implementation.
+        let sq_norm = encoded_gradient
+            .chunks(self.cfg.chunk_len_sq_norm_equal)
+            .map(|chunk| {
+                for x in chunk {
+                    buf.push(*x);
```

Comment:
double-check: push twice?

Reply:
Yup, this is correct! `buf` is a sequence of inputs to the `Mul` gadget, which has two input wires and outputs the multiplication of the inputs. Here we're just using the `Mul` gadget to square each of the gradient entries.